aten/src/ATen/native/cpu/ReduceNormKernel.cpp

```diff
@@ -0,0 +1,138 @@
+  } else if (val == static_cast<double>(1)) {
+    binary_kernel_reduce(iter, NormOneOps<scalar_t, acc_t, out_t>(), acc_t(0));
+  } else if (val == static_cast<double>(2)) {
+    binary_kernel_reduce(iter, NormTwoOps<scalar_t, acc_t, out_t>(), acc_t(0));
+  } else if (val == static_cast<double>(INFINITY)) {
+    binary_kernel_reduce(iter, AbsMaxOps<scalar_t, acc_t, out_t>(), acc_t(0));
+  } else if (val == static_cast<double>(-INFINITY)) {
+    binary_kernel_reduce(
+        iter,
+        AbsMinOps<scalar_t, acc_t, out_t>(),
+        std::numeric_limits<acc_t>::infinity());
+  } else {
+    binary_kernel_reduce(
+        iter, NormOps<scalar_t, acc_t, out_t>{acc_t(val)}, acc_t(0));
+  }
+}
+
+void norm_kernel_cpu(TensorIterator& iter, const Scalar& p) {
+  double val;
+  if (p.isIntegral(false)) {
+    val = p.to<int64_t>();
+  } else if (p.isFloatingPoint()) {
+    val = p.to<double>();
+  } else {
+    TORCH_CHECK(false, "norm_kernel_cpu expects norm to be integer or float");
+  }
+  if (iter.numel() == 0) {
+    iter.output().fill_((val < 0) ? INFINITY : 0);
+    return;
+  }
+
+  if (val == static_cast<double>(2) && is_reduce_lastdim(iter) &&
+      iter.dtype(0) == iter.input_dtype() &&
+      (iter.input_dtype() == kFloat || iter.input_dtype() == kDouble ||
+       iter.input_dtype() == kBFloat16)) {
+    AT_DISPATCH_FLOATING_TYPES_AND(
+        kBFloat16, iter.input_dtype(), "norm_cpu", [&] {
+          // use float as accumulate type for BFloat16
+          using acc_t = at::opmath_type<scalar_t>;
+          binary_kernel_reduce_lastdim(
+              iter,
+              [](char* result_data_bytes, char* self_data_bytes, int64_t size) {
+                scalar_t* result_data = (scalar_t*)result_data_bytes;
+                scalar_t* self_data = (scalar_t*)self_data_bytes;
+
+                using Vec = Vectorized<scalar_t>;
+                using fVec = Vectorized<acc_t>;
+                fVec acc_vec{acc_t(0)};
+                acc_t buffer[fVec::size()];
+                int64_t d = 0;
+                for (; d < size - (size % Vec::size()); d += Vec::size()) {
+                  Vec data_vec = Vec::loadu(self_data + d);
+                  norm_two_reduce_step(acc_vec, data_vec);
+                }
+                acc_vec.store(buffer);
+                for (int j = 1; j < fVec::size(); j++) {
+                  buffer[0] = buffer[0] + buffer[j];
+                }
+                for (; d < size; d++) {
+                  acc_t data_val = acc_t(self_data[d]);
+                  buffer[0] += data_val * data_val;
+                }
+                result_data[0] = scalar_t(std::sqrt(buffer[0]));
+              });
+        });
+  } else {
+    if (iter.dtype(0) == kHalf) {
+      return norm_kernel_cpu_impl<at::Half, float>(iter, val);
+    } else if (iter.dtype(0) == kBFloat16) {
+      return norm_kernel_cpu_impl<at::BFloat16, float>(iter, val);
+    }
```

Comment:
You can copy the logic for how CUDA does it, as it's able to fuse this operation and a casting operation in one in some cases.

Reply:
Yes, using the same logic as cuda.